File: flask_app/app/models/order.py
from app.models.base import BaseModel, db
from app.constants import OrderStatus


class Order(BaseModel):
    __tablename__ = "order"

    user_id = db.Column(db.String(36), nullable=False)
    status = db.Column(
        db.Enum(OrderStatus), default=OrderStatus.PENDING, nullable=False
    )
    total_amount = db.Column(db.Float, nullable=False, default=0.0)
    total_savings = db.Column(db.Float, nullable=False, default=0.0)
    subtotal = db.Column(db.Float, nullable=False, default=0.0)

    delivery_address = db.Column(db.Text, nullable=False)
    delivery_fee = db.Column(db.Float, nullable=False, default=0.0)
    delivery_fee_applied = db.Column(db.Boolean, nullable=False, default=True)

    items = db.Column(db.JSON, nullable=False, default=[])
    discounts = db.Column(db.JSON, nullable=False, default=[])

    @property
    def order_item_count(self):
        return len(self.items) if self.items else 0

    def __repr__(self):
        return f"<Order {self.id}>"

File: java_membership_api/src/main/java/com/example/membership_api/model/Order.java
package com.example.membership_api.model;

import java.util.List;
import java.util.Map;

import com.example.membership_api.constants.OrderStatus;
import com.example.membership_api.utils.JsonUtils;
import com.fasterxml.jackson.core.type.TypeReference;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Table;
import lombok.Data;
import lombok.EqualsAndHashCode;

@Data
@Entity
@Table(name = "orders")
@EqualsAndHashCode(callSuper = true)
public class Order extends BaseModel {

    @Column(nullable = false, length = 36)
    private String userId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderStatus status = OrderStatus.PENDING;

    @Column(nullable = false)
    private double totalAmount = 0.0;

    @Column(nullable = false)
    private double totalSavings = 0.0;

    @Column(nullable = false)
    private double subtotal = 0.0;

    @Column(nullable = false)
    private String deliveryAddress;

    @Column(nullable = false)
    private double deliveryFee = 0.0;

    @Column(nullable = false)
    private boolean deliveryFeeApplied = true;

    @Column(columnDefinition = "JSON", nullable = false)
    private String items="[]";  

    @Column(columnDefinition = "JSON", nullable = false)
    private String discounts="[]"; 

    // Helper method
    public int getOrderItemCount() {
        try {
            List<String> itemList = JsonUtils.parseJson(items, new TypeReference<List<String>>() {});
            // List<Map<String, Object>> itemList = JsonUtils.parseJson(items);
            System.out.println("Parsed item count: " + itemList.size());
            return itemList.size();
        } catch (Exception e) {
            System.err.println("Failed to parse items: " + e.getMessage());
            return 0;
        }
    }
}
